include/zephyr/drivers/emul.h

```diff
@@ -67,30 +69,78 @@ struct emul {
 extern const struct emul __emul_list_start[];
 extern const struct emul __emul_list_end[];
 
-/* Use the devicetree node identifier as a unique name. */
+/**
+ * @brief Use the devicetree node identifier as a unique name.
+ */
+#define EMUL_DT_NAME_GET(node_id) (_CONCAT(__emulreg_, node_id))
+
+/**
+ * @brief Use the devicetree node identifier as a unique name.
+ *
+ * @deprecated Use EMUL_DT_NAME_GET instead
+ */
 #define EMUL_REG_NAME(node_id) (_CONCAT(__emulreg_, node_id))
 
 /**
- * Define a new emulator
+ * @brief Define a new emulator
  *
  * This adds a new struct emul to the linker list of emulations. This is
  * typically used in your emulator's DT_INST_FOREACH_STATUS_OKAY() clause.
  *
- * @param init_ptr function to call to initialise the emulator (see emul_init
- *	typedef)
  * @param node_id Node ID of the driver to emulate (e.g. DT_DRV_INST(n))
- * @param cfg_ptr emulator-specific configuration data
+ * @param init_fn function to call to initialise the emulator (see emul_init
+ *	typedef)
  * @param data_ptr emulator-specific data
+ * @param cfg_ptr emulator-specific configuration data
  */
-#define EMUL_DEFINE(init_ptr, node_id, cfg_ptr, data_ptr)                                          \
-	static struct emul EMUL_REG_NAME(node_id) __attribute__((__section__(".emulators")))       \
+#define EMUL_DT_DEFINE(node_id, init_fn, data_ptr, cfg_ptr)                                        \
+	const struct emul EMUL_DT_NAME_GET(node_id) __attribute__((__section__(".emulators")))     \
 	__used = {                                                                                 \
-		.init = (init_ptr),                                                                \
+		.init = (init_fn),                                                                 \
 		.dev_label = DT_LABEL(node_id),                                                    \
 		.cfg = (cfg_ptr),                                                                  \
 		.data = (data_ptr),                                                                \
 	};
 
+/**
+ * @brief Define a new emulator.
+ *
+ * @deprecated Use EMUL_DT_DEFINE or EMUL_DT_INST_DEFINE.
+ */
+#define EMUL_DEFINE(init_ptr, node_id, cfg_ptr, data_ptr)                                          \
+	EMUL_DT_DEFINE(node_id, init_ptr, data_ptr, cfg_ptr)
```

Comment:
Heh, `#define X(a, b, c, d) Y(b, a, d, c)` seems maybe a little cruel for users of the earlier API, but then I guess it's being deprecated and the new argument order looks more conventional.